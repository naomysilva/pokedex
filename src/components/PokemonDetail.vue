<template>
  <div class="detail">
    <div class="detail-view-wrapper" v-if="show">
      <div class="detail-view" >
        <div v-if="pokemon" class="image">
          <img :src="imageUrl + pokemon.id + '.png'" alt="" />
        </div>
        <div v-if="pokemon" class="data">
          <h2>{{ pokemon.name }}</h2>
          <div class="property">
            <div class="left">Base Experience</div>
            <div class="right">{{ pokemon.base_experience }} XP</div>
          </div>
          <div class="property">
            <div class="left">Height</div>
            <div class="right">{{ pokemon.height / 10 }} m</div>
          </div>
          <div class="property">
            <div class="left">Weight</div>
            <div class="right">{{ pokemon.weight / 10 }} kg</div>
          </div>

          <h3>Pokemon game_indices</h3>
          <div class="types">
            <div
              class="type"
              v-for="(value, index) in pokemon.game_indices"
              :key="'value' + index"
            >
              {{ value.game_index }}
              {{ value.version.name }}
            </div>
          </div>
          <h3>Pokemon Moves</h3>
          <div class="moves">
            <div
              class="move"
              v-for="(value, index) in pokemon.stats"
              :key="'stats' + index"
            >
              {{ value.stat.name }} = {{ value.base_stat }}
            </div>
          </div>

          <h3>Abilities</h3>
          <div class="abilities">
            <div
              class="ability"
              v-for="(value, index) in pokemon.abilities"
              :key="'value' + index"
            >
              {{ value.ability.name }}
            </div>
          </div>

          <h3>
            moves do Pokémon
            <div class="sprites row">
              <div
                class="sprite col-md-3"
                v-for="(value, key) in pokemon.sprites"
                :key="key"
              >
                <img
                  v-if="typeof value !== 'object' && value !== null"
                  :src="value"
                  :alt="key"
                />
              </div>
            </div>
          </h3>
          <h3>Moves</h3>
          <div class="moves">
            <div class="move" v-for="(move, index) in pokemon.moves" :key="index">
              <div v-if="move.move && move.move.name">
                {{ move.move.name }}
              </div>
            </div>
          </div>

          <h3>Other</h3>
          <div class="sprites row">
            <div
              class="sprite col-md-2"
              v-for="(value, key) in pokemon.sprites.other"
              :key="key"
            >
              <img
                v-if="value !== null"
                :src="value.front_default"
                :alt="key"
                height="75px"
              />
            </div>
          </div>
          <h3>Gerações</h3>

          <div class="sprites row">
            <div
              v-for="(generation, generationName) in pokemon.sprites.versions"
              :key="generationName"
            >
              <h2>{{ generationName }}</h2>
              <div
                v-for="(version, versionName) in generation"
                :key="versionName"
                class="row"
              >
                <div v-for="(value, key) in version" :key="key" class="col-md-3">
                  <img
                    v-if="typeof value === 'string'"
                    :src="value"
                    :alt="key"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <h2 v-else>The pokemon was not found</h2>
        <button class="close" @click="closeDetail">close</button>
      </div>
    </div>
    <i v-else class="fas fa-spinner fa-spin"></i>
  </div>
</template>

<script>
export default {
  props: ["pokemonUrl", "imageUrl"],
  data() {
    return {
      show: false,
      pokemon: {},
    };
  },
  methods: {
    fetchData() {
      let req = new Request(this.pokemonUrl);
      fetch(req)
        .then((resp) => {
          if (resp.status === 200) return resp.json();
        })
        .then((data) => {
          this.pokemon = data;
          this.show = true;
        })
        .catch((error) => {
          console.log(error);
        });
    },
    closeDetail() {
      this.$emit("closeDetail");
    },
  },
  created() {
    this.fetchData();
  },
};
</script>

<style scoped>
.detail {
  display: flex;
  justify-content: center;
  align-items: flex-start;
  position: fixed;
  top: 0;
  left: 0;
  padding: 90px 10px 10px;
  overflow-y: hidden;
}

.detail-view-wrapper {
  max-height: calc(100vh - 100px);
  overflow-y: auto; 
}

.detail-view {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  position: relative;
  width: 100%;
  max-width: 510px;
  padding: 50px 0 0;
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 0 15px 30px rgba(0, 0, 0, 0.2), 0 10px 10px rgba(0, 0, 0, 0.2);
  margin-top: 6em;
}

.image {
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  top: -60px;
  width: 120px;
  height: 120px;
  background-color: #333;
  border-radius: 50%;
  overflow: hidden;
  box-shadow: 0 15px 30px rgba(0, 0, 0, 0.2), 0 10px 10px rgba(0, 0, 0, 0.2);
}

h2 {
  text-transform: capitalize;
}

.data {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  flex-direction: column;
  width: 100%;
  margin-bottom: 40px;
}

.property {
  width: 90%;
  max-width: 400px;
  border-bottom: 1px solid #ccc;
  margin-bottom: 10px;
}

.left {
  float: left;
}
.right {
  float: right;
}

h3 {
  width: 90%;
  max-width: 400px;
  border-bottom: 1px solid #ccc;
}

.types,
.abilities, .moves {
  display: flex;
  justify-content: flex-start;
  flex-wrap: wrap;
  width: 90%;
  max-width: 400px;
}
.type,
.ability,.move {
  margin: 0 10px 10px 0;
  padding: 5px 10px;
  border-radius: 20px;
  color: #fff;
  font-size: 1rem;
  letter-spacing: 2px;
  text-transform: capitalize;
  word-wrap: none;
  word-break: keep-all;
}

.type, .move {
  background-color: #0a2e50;
}
.ability {
  background-color: #c73015;
}

.close {
  outline: none;
  border: none;
  border-radius: 5px;
  background-color: #333;
  color: #efefef;
  padding: 10px 20px;
  margin-bottom: 20px;
  font-size: 1.2rem;
  cursor: pointer;
}

i {
  font-size: 2rem;
  color: #efefef;
}
</style>
