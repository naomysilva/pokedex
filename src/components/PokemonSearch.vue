<template>
  <div class="container">
    <div class="row justify-content-center">
      <div class="col-12 col-md-6">
        <div class="searchbar">
          <form @submit.prevent="setPokemonUrl" class="d-flex justify-content-center align-items-center">
            <input type="text" v-model="searchvalue" class="form-control" />
            <button class="btn btn-primary mx-3" @click.prevent="setPokemonUrl">Buscar</button>
          </form>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  props: ["apiUrl"],
  data() {
    return {
      searchvalue: "",
    };
  },
  methods: {
    setPokemonUrl() {
      if (this.searchvalue !== "")
        this.$emit("setPokemonUrl", this.apiUrl + this.searchvalue);
    },
  },
};
</script>

<style scoped>
.searchbar {
  padding-bottom: 20px;
}
input {
  border-radius: 5px;
  padding: 10px;
  font-size: 1rem;
}

i {
  font-size: 1.25rem;
  color: #0a2e50;
  cursor: pointer;
}
</style>
