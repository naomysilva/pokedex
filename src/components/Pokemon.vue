<template>
  <div class="container-fluid">
  <div class="row">
    <h1 class="text-center pt-4">Pokemon's</h1>
    <PokemonSearch 
      :apiUrl="apiUrl" 
      @setPokemonUrl="setPokemonUrl" />
    <PokemonList 
      :imageUrl="imageUrl" 
      :apiUrl="apiUrl"
      @setPokemonUrl="setPokemonUrl" />
    <PokemonDetail 
      v-if="showDetail"
      :pokemonUrl="pokemonUrl"
      :imageUrl="imageUrl"
      @closeDetail="closeDetail" />
  </div>
</div>
</template>

<script>
  import PokemonSearch from './PokemonSearch.vue';
  import PokemonList from './PokemonList.vue';
  import PokemonDetail from './PokemonDetail.vue';

  export default {
    data: () => {
      return {
        imageUrl: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/',
        apiUrl: 'https://pokeapi.co/api/v2/pokemon/',
        pokemonUrl: '',
        showDetail: false
      }
    },
    components: {
      PokemonSearch,
      PokemonList,
      PokemonDetail
    },
    methods: {
      setPokemonUrl(url) {
        this.pokemonUrl = url;
        this.showDetail = true;
      },
      closeDetail() {
        this.pokemonUrl = '';
        this.showDetail = false;
      }
    }
  }
</script>

<style scoped>
  @import url('https://fonts.googleapis.com/css?family=Acme');

  .container-fluid {
    background: radial-gradient(#156F99, #0A2E50);

    font-family: 'Acme', arial;
    font-size: 1rem;
    font-weight: normal;
  }

  h1 { color: #efefef; }
</style>