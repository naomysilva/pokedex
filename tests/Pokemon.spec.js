import { shallowMount } from '@vue/test-utils';
import Pokemon from '@/components/Pokemon.vue';
import PokemonSearch from '@/components/PokemonSearch.vue';
import PokemonDetail from '@/components/PokemonDetail.vue';

describe('Pokemon.vue', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(Pokemon);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('sets Pokemon URL when PokemonSearch emits setPokemonUrl event', () => {
    const wrapper = shallowMount(Pokemon);
    const searchComponent = wrapper.findComponent(PokemonSearch);
    searchComponent.vm.$emit('setPokemonUrl', 'https://pokeapi.co/api/v2/pokemon/1');
    expect(wrapper.vm.pokemonUrl).toBe('https://pokeapi.co/api/v2/pokemon/1');
    expect(wrapper.vm.showDetail).toBe(true);
  });
});
