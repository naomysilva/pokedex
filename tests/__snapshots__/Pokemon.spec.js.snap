// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon.vue renders correctly 1`] = `
<div
  class="container"
>
  <h1>
    Pokemon's
  </h1>
  <pokemon-search-stub
    apiurl="https://pokeapi.co/api/v2/pokemon/"
  />
  <pokemon-list-stub
    apiurl="https://pokeapi.co/api/v2/pokemon/"
    imageurl="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"
  />
  <!--v-if-->
</div>
`;
